member_map.go

```diff
@@ -17,66 +17,215 @@
 package swim
 
 import (
+	"errors"
 	"net"
 	"strconv"
 	"sync"
 	"time"
+
+	"math/rand"
 )
 
-// status of members
+var ErrEmptyMemberID = errors.New("MemberID is empty")
+
+// Status of members
 type Status int
 
 const (
 
-	// unknown status of a member
+	// Unknown status of a member
 	Unknown Status = iota
 
-	// alive status
+	// Alive status
 	Alive
 
-	// suspicious status of whether a member is dead or not
+	// Suspicious status of whether a member is dead or not
 	Suspected
 
-	// dead status
+	// Dead status
 	Dead
 )
 
 type MemberID struct {
 	ID string
 }
 
-// struct of Member
+// Struct of Member
 type Member struct {
 
-	// ip address
+	// Id of member
+	ID MemberID
+
+	// Ip address
 	Addr net.IP
 
-	// port
+	// Port
 	Port uint16
 
 	// Current member status from my point of view
-	status Status
+	Status Status
 
 	// Time last status change happened
-	lastStatusChange time.Time
+	LastStatusChange time.Time
 
 	// Incarnation helps to keep the most fresh information about member status in the system
 	// which tells that suspect member confirming that it is alive, and only when suspect
 	// got suspicion message, that member can increments incarnation
-	incarnation uint32
+	Incarnation uint32
 }
 
-// convert member addr and port to string
+// Convert member addr and port to string
 func (m *Member) Address() string {
 	return net.JoinHostPort(m.Addr.String(), strconv.Itoa(int(m.Port)))
 }
 
+// Get
+func (m *Member) GetID() MemberID {
+	return m.ID
+}
+
 type MemberMap struct {
-	sync.RWMutex
-	members map[MemberID]*Member
+	lock    sync.RWMutex
+	members map[MemberID]Member
+
+	// This is for selecting k random member based on round-robin
+	waitingMembers []MemberID
+}
+
+func NewMemberMap() *MemberMap {
+	return &MemberMap{
+		members:        make(map[MemberID]Member),
+		waitingMembers: make([]MemberID, 0),
+		lock:           sync.RWMutex{},
+	}
 }
 
-// select K random member from members
-func (m *MemberMap) selectKRandomMember(k int) []Member {
+// Select K random memberID from waitingMembers(length of returning member can be lower than k).
+// ** WaitingMembers are shuffled every time when members are updated **, so just returning first K item in waitingMembers is same as
```

Comment:
> ** WaitingMembers are shuffled every time when members are updated **

I think this comment will be more precise 

> ** WaitingMembers are shuffled every time when waiting members are updated **

I think there's no logic updating **waiting members** when **members** are updated

Reply:
`m.waitingMembers = resetWaitingMembersID(m.members)` this function shuffles waiting members using members. `m.waitingMembers = resetWaitingMembersID(m.members)` located in the lower part of addmember function

Reply:
yes `resetWatingMembersID(m.members)` shuffles waiting members using *members* but I think there's no **when members are updated** part in the function